### PR TITLE
Parametrize hostname extraction logic used in oidc filter to set authentication cookies

### PR DESCRIPTION
**Pull request by abinet**
The PR addresses the issue
https://github.com/zalando/skipper/issues/1909

### CODE REVIEW

File: filters/auth/oidc.go [hunk 1]
```diff
@@ -165,6 +168,21 @@ func (s *tokenOidcSpec) CreateFilter(args []interface{}) (filters.Filter, error)
...
+		if err != nil {
+			return nil, err
+		}
+		if subdomainsToRemove < 0 {
+			return nil, fmt.Errorf("domain level cannot be negative '%s'", sargs[paramSubdomainsToRemove])
+		}
+
+		if err != nil || sargs[paramCallbackURL] == "" {
+			return nil, fmt.Errorf("invalid redirect url '%s': %v", sargs[paramCallbackURL], err)
+		}
```
Comment (AlexanderYastrebov) [inline filters/auth/oidc.go:183]: This is redundant

File: filters/auth/oidc.go [hunk 2]
```diff
@@ -165,6 +168,21 @@ func (s *tokenOidcSpec) CreateFilter(args []interface{}) (filters.Filter, error)
 		return nil, err
 	}
 
+	subdomainsToRemove := 1
+	if len(sargs) > paramSubdomainsToRemove && sargs[paramSubdomainsToRemove] != "" {
```
Comment (AlexanderYastrebov) [inline filters/auth/oidc.go:172]: Looking at https://github.com/zalando/skipper/blob/7e92c9ab9bd7c791e4388bfcc5f73b71cee3751f/filters/auth/oidc.go#L430-L438

I am wondering if we should rather introduce `Cookie Options` similar to `Auth Code Options` to allow multiple cookie-related options, e.g. to customize path, name or domain.

File: filters/auth/oidc_test.go [hunk 3]
```diff
@@ -549,6 +554,87 @@ func TestCreateFilterOIDC(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestCreateFilterOIDC_subdomainsToRemove(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (abinet) [reply filters/auth/oidc_test.go:559]: TestOIDCSetup has already huge complexity. Adding domain setting verification into this test case will increase complexity even more. We only have to test whether parameter is set properly and you use same approach with the Test_tokenOidcFilter_getMaxAge already. That's the reasons why I introduced a new test case. However if you still prefer to bloat TestOIDCSetup  I will do this.

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:559]: > you use same approach with the Test_tokenOidcFilter_getMaxAge already
Cookie age is verified in `TestOIDCSetup` though. 

We need to verify that cookie has an expected domain and we do not need to cover all the cases maybe as `TestExtractDomainFromHost` does that already.

File: docs/reference/filters.md [hunk 4]
```diff
@@ -1378,6 +1378,7 @@ The filter needs the following parameters:
...
 * **Auth Code Options** (optional) Passes key/value parameters to a provider's authorization endpoint. The value can be dynamically set by a query parameter with the same key name if the placeholder `skipper-request-query` is used.
 * **Upstream Headers** (optional) The upstream endpoint will receive these headers which values are parsed from the OIDC information. The header definition can be one or more header-query pairs, space delimited. The query syntax is [GJSON](https://github.com/tidwall/gjson/blob/master/SYNTAX.md).
+* **SubdomainsToRemove** (optional) The setting configures hostname used to store oidc cookie. If skipper is being used in front of multiple domenvironments configured on subdomains one can store cookie to a parent domain in order to keep authentication info for all the environments. The non-negative integer configures a number of subdomain levels which must be removed from initial hostname. If no value set one subdomain will be removed from hostnames containing more then 2 domain levels (www.example.com -> example.com). Set 0 to skip this behaviour and not remove anything.
```
Comment (AlexanderYastrebov) [inline docs/reference/filters.md:1381]: I would rephrase a bit to simplify

* **SubdomainsToRemove** (optional, default "1") Configures number of subdomains to remove from the request hostname to derive OIDC cookie domain. By default one subdomain is removed, e.g. for the `www.example.com` request hostname the OIDC cookie domain will be `example.com`. Configure "0" to use the same hostname. Note that value is a string.

File: filters/auth/oidc.go [hunk 5]
```diff
@@ -368,10 +387,10 @@ func extractDomainFromHost(host string) string {
 	if ip != nil {
 		return ip.String()
 	}
-	if strings.Count(h, ".") < 2 {
+	if 1+strings.Count(h, ".")-subdomainsToRemove < 2 {
 		return h
 	}
-	return strings.Join(strings.Split(h, ".")[1:], ".")
+	return strings.Join(strings.Split(h, ".")[subdomainsToRemove:], ".")
```
Comment (AlexanderYastrebov) [inline filters/auth/oidc.go:393]: There is no need split if there nothing to remove and there is no need to count if we split anyway:
```go
if subdomainsToRemove == 0 {
    return h
}
subDomains := strings.Split(h, ".")
if len(subDomains) - subdomainsToRemove < 2 {
    return h
}
return strings.Join(subDomains[subdomainsToRemove:], ".")
```

File: filters/auth/oidc_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (abinet) [reply filters/auth/oidc_test.go:559]: Just tried to implement a check in TestOIDCSetup, however I am afraid it is not possible to provide a custom hostname for the callback call instead of ip address. Any suggestions?

[the remark this answers is not in the document]
Comment (szuecs) [reply filters/auth/oidc_test.go:559]: Maybe you can try to use `localhost` as TLD.
On my machine this works and could be used for testing (replace host part of the listener URL if there is a configured domain for the purpose:
```
% host foo.bar.localhost                                                                                                                                                                dev
foo.bar.localhost has address 127.0.0.1
foo.bar.localhost has IPv6 address ::1
```

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:559]: > Just tried to implement a check in TestOIDCSetup, however I am afraid it is not possible to provide a custom hostname for the callback call instead of ip address. Any suggestions?

It is possible but the problem is that it should resolve to the test server address (e.g. 127.0.0.1) and the only one that is universal is `localhost` but it does not support subdomains afaik.

> Maybe you can try to use localhost as TLD.
> On my machine this works

Works on Ubuntu as well but afaik `localhost` can not have subdomains https://stackoverflow.com/questions/19016553/add-subdomain-to-localhost-url so maybe it is due to `systemd-resolved` https://unix.stackexchange.com/questions/401966/how-does-every-subdomain-of-localhost-point-to-localhost-on-fedora

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:559]: It is possible like so
```diff
diff --git a/filters/auth/oidc_test.go b/filters/auth/oidc_test.go
index a7e505b6..27a5c613 100644
--- a/filters/auth/oidc_test.go
+++ b/filters/auth/oidc_test.go
@@ -731,7 +731,11 @@ func TestOIDCSetup(t *testing.T) {
                                t.Errorf("Failed to parse url %s: %v", proxy.URL, err)
                        }
 
-                       oidcServer := createOIDCServer(proxy.URL+"/redirect", tc.client, tc.clientsecret)
+                       redirectURL, _ := url.Parse(proxy.URL)
+                       redirectURL.Host = "localhost:" + redirectURL.Port()
+                       redirectURL.Path = "/redirect"
+
+                       oidcServer := createOIDCServer(redirectURL.String(), tc.client, tc.clientsecret)
                        defer oidcServer.Close()
                        t.Logf("oidc/auth server URL: %s", oidcServer.URL)
 
@@ -739,7 +743,7 @@ func TestOIDCSetup(t *testing.T) {
                        sargs := []interface{}{
                                tc.client,
                                tc.clientsecret,
-                               proxy.URL + "/redirect",
+                               redirectURL.String(),
                        }
 
                        // test that, we get an error if provider is no url
@@ -827,7 +831,7 @@ func TestOIDCSetup(t *testing.T) {
                        t.Logf("Got body: %s", bs)
 
                        // make sure cookies have correct maxAge according to exp
-                       for _, c := range client.Jar.Cookies(reqURL) {
+                       for _, c := range client.Jar.Cookies(redirectURL) {
                                maxAge := time.Duration(c.MaxAge) * time.Second
                                assert.True(t, maxAge >= 2*time.Hour-time.Minute && maxAge <= 2*time.Hour,
                                        "maxAge has to be within [2h - 1m, 2h]")
```

It also seems that cookie testing is not really strict, e.g. there is a test case where cookies are not expected and test does not check that, also it uses `reqURL` but it only works because redirect url host is IP address (or localhost in above) and hence cookie domain is IP (or localhost) as well.

The proper cookie testing should expect (or expect no) cookies and check domain which in case of multi-level hostname would not be equal to the `reqURL` or `redirectURL` due to subdomain removal.

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:559]: I guess we could hack http clients (also OAuth2 client) to resolve any desired domain to 127.0.0.1 but that needs thinking and refactoring.
@szuecs  Should we leave cookie domain testing out of scope here until we find a way to do it properly, wdyt?

[the remark this answers is not in the document]
Comment (szuecs) [reply filters/auth/oidc_test.go:559]: I am fine with both. So either check it in the client, but this needs a custom Dial() or DialContext() in the http.Transport as far as I remember or omit this part in testing and create an issue to postpone it.

File: filters/auth/oidc.go (hunk 2 above)
Comment (szuecs) [reply filters/auth/oidc.go:172]: sgtm

File: filters/auth/oidc_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (szuecs) [reply filters/auth/oidc_test.go:559]: I guess you can always make it if the client supports reading `/etc/hosts`, but systemd-resolved seems to be ok for automated tests. I would just try it and see what GH actions (and for maintainers CDP) will tell us about.

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:559]: I have opened an issue https://github.com/zalando/skipper/issues/1926 and will have a look into it.

[the remark this answers is not in the document]
Comment (szuecs) [reply filters/auth/oidc_test.go:559]: thanks!

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:559]: The https://github.com/zalando/skipper/pull/1927 was merged and now it should be possible to test cookie domain. Please rebase to see example tests.

**Comment by szuecs** [comment]
@abinet can you rebase on master?
Alexander added the domain test and you can use it in your PR.

**Comment by abinet** [comment]
Thank you @AlexanderYastrebov  
@szuecs rebased as requested.

Now the TestOIDCSetup validates default behaviour.
The rest is being tested by TestExtractDomainFromHost.

Let me know if we need more actions on this PR

File: docs/reference/filters.md [hunk 6]
```diff
@@ -1381,6 +1381,7 @@ The filter needs the following parameters:
 * **Claims** The claims which should be present in the token returned by the provider.
 * **Auth Code Options** (optional) Passes key/value parameters to a provider's authorization endpoint. The value can be dynamically set by a query parameter with the same key name if the placeholder `skipper-request-query` is used.
 * **Upstream Headers** (optional) The upstream endpoint will receive these headers which values are parsed from the OIDC information. The header definition can be one or more header-query pairs, space delimited. The query syntax is [GJSON](https://github.com/tidwall/gjson/blob/master/SYNTAX.md).
+* **SubdomainsToRemove** (optional, default "1") Configures number of subdomains to remove from the request hostname to derive OIDC cookie domain. By default one subdomain is removed, e.g. for the www.example.com request hostname the OIDC cookie domain will be example.com. Configure "0" to use the same hostname. Note that value is a string.
```
Comment (szuecs) [inline docs/reference/filters.md:1384]: Maybe add that the default choice is used to enable SSO across all UI applications of a domain `example.com`.

**Comment by szuecs** [comment]
:+1:

File: filters/auth/oidc_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply filters/auth/oidc_test.go:562]: This test creates createOIDCServer that is not used.

Lets remove this test and add a test or two to the `TestOIDCSetup` integration test because we are still not testing domain handling when new option is different from the default.

File: filters/auth/oidc.go (hunk 2 above)
Comment (AlexanderYastrebov) [reply filters/auth/oidc.go:172]: @szuecs Lets decide if we go for scalar parameter like proposed and add more cookie-related parameters if needed in the future or we go for extensible key-value parameters for cookie config similar to auth code options.

Comment (szuecs) [reply filters/auth/oidc.go:172]: I gue a map is not too complicated but I don't know what value this has. It depends on the dns setup but we can argue a map will add more freedom to the user, but it is also more complicated.
I don't know. What is your favorite?

**Comment by abinet** [comment]
@AlexanderYastrebov I've moved all the tests to TestOIDCSetup as requested. Please review

**Comment by szuecs** [comment]
:+1:

File: filters/auth/oidc_test.go [hunk 7]
```diff
@@ -803,8 +844,10 @@ func TestOIDCSetup(t *testing.T) {
 				reqURL.RawQuery = q.Encode()
 			}
 
-			if tc.upstreamheaders != "" {
-				sargs = append(sargs, tc.upstreamheaders)
+			sargs = append(sargs, tc.upstreamheaders)
```
Comment (AlexanderYastrebov) [inline filters/auth/oidc_test.go:847]: This will append empty value unconditionally. The code seems to check for empty value so this would work. 

For the consistency lets remove `if tc.subdomainsToRemove != ""` check as well as your code also checks for empty value.

File: filters/auth/oidc.go (hunk 2 above)
Comment (AlexanderYastrebov) [reply filters/auth/oidc.go:172]: > What is your favorite?

Lets not overcomplicate then and have it like proposed. The oidc* filter interface is bloated already we can not improve it much anyway.

**Comment by AlexanderYastrebov** [comment]
:+1:
